utils/logger.go

```diff
@@ -0,0 +1,109 @@
+// this is a new logger interface for mattermost
+
+package utils
+
+import (
+	"context"
+	"encoding/json"
+	"runtime"
+
+	l4g "github.com/alecthomas/log4go"
+)
+
+// this pattern allows us to "mock" the underlying l4g code when unit testing
+var debug = l4g.Debug
+var info = l4g.Info
+var err = l4g.Error
+
+// Logger - an instance of the log framework that can be used to log messages
+type Logger string
+
+// NewLogger - creates a new instance of Logger named after the file that created it
+func NewLogger() Logger {
+	_, file, _, ok := runtime.Caller(1)
+	if ok {
+		return Logger(file)
+	}
+	return Logger("Unknown Logger")
+}
+
+// contextKey lets us add contextual information to log messages
+type contextKey string
+
+func (c contextKey) String() string {
+	return string(c)
+}
+
+const contextKeyUserID contextKey = contextKey("user-id")
```

Comment:
The custom type is the suggested pattern when working with Go's Context object. Check out this blog post for a better explanation than I can provide: https://blog.golang.org/context

As for the style note, thank you, I'll keep that in mind. I've been using VSCode, and it is rather opinionated about code style. It was quite insistent that I capitalize I and D. I'll have to figure out a way to shut it up.